src/ssd/ssd-titlebar.c

```diff
@@ -241,6 +244,58 @@ set_alt_button_icon(struct ssd *ssd, enum ssd_part_type type, bool enable)
 	} FOR_EACH_END
 }
 
+/*
+ * Usually this function sets ssd->state.visible_buttons_count_{left,right} to
+ * wl_list_length(&rc.title_buttons_{left,right}), but smaller numbers can be
+ * set for tiny windows.
+ */
+static void
+update_visible_buttons(struct ssd *ssd)
+{
+	struct view *view = ssd->view;
+	int width = view->current.width;
+	int button_width = view->server->theme->window_button_width;
+	int button_count_left = wl_list_length(&rc.title_buttons_left);
+	int button_count_right = wl_list_length(&rc.title_buttons_right);
+
+	/*
+	 * The corner-left button is lastly removed as it's usually a window
+	 * menu or app icon button.
+	 */
+	while (width < button_width * (button_count_left + button_count_right)) {
+		if (button_count_left > button_count_right) {
+			button_count_left--;
+		} else {
+			button_count_right--;
+		}
+	}
```

Comment:
Here's a deterministic algorithm that can be executed without loops and should preserve the behavior you desire:
```c
int max_buttons = width / button_width; 
int lb = MIN(button_count_left, (max_buttons + 1) / 2);
int rb = MIN(button_count_right, max_buttons / 2);
button_count_left = MIN(button_count_left, max_buttons - rb);
button_count_right = MIN(button_count_right, max_buttons - button_count_left);
```
I've enumerated all possible combinations of (left, right) configurations in our six-button set and confirmed that the removals are approximately uniform, while ensuring that the first left button does not disappear unless it was never configured in the first place, or the number of buttons allowed is identically zero.

Reply:
Well, I usually don't like math and I prefer the simplest code as long as it doesn't have a noticeable overhead.

My code makes it clear that `button_count_{left,right}` are just `wl_list_length(&rc.title_buttons_{left,right})` when the view width is large enough and how button counts are decremented as the view shrinks.

Reply:
If you want to keep the loop, you might consider setting

```c
int button_width = MAX(0, view->server->theme->window_button_width);
```

to guard against some invalid value inadvertently triggering an infinite loop.

Reply:
Good idea. I added `assert(button_width > 0)` as negative values are already checked when reading `themerc`:

https://github.com/labwc/labwc/blob/b34f2c9d85a52bc19a9458b6d38bae64fae55baa/src/theme.c#L740-L747